pkg/streamer/file.go

```diff
@@ -194,128 +195,61 @@ func fileSizeUpdated(path string, latestSize int64) (int, error) {
+	for {
+		select {
+		case <-ctx.Done():
+			if err := ctx.Err(); err != nil {
+				errCh <- err
+			}
+			return
+		case <-ticker.C:
+			// check the latest relay log file whether updated when adding watching and collecting newer
+			cmp, err := fileSizeUpdated(latestFilePath, latestFileSize)
+			switch {
+			case err != nil:
+				errCh <- terror.Annotatef(err, "latestFilePath=%s latestFileSize= %d", latestFilePath, latestFileSize)
 				return
-			case err2, ok := <-watcher2.Errors:
-				if !ok {
-					result <- watchResult{
-						updatePath: "",
-						err:        terror.ErrWatcherChanClosed.Generate("errors", dir),
-					}
-				} else {
-					result <- watchResult{
-						updatePath: "",
-						err:        terror.ErrWatcherChanRecvError.Delegate(err2, dir),
-					}
-				}
+			case cmp < 0:
+				errCh <- terror.ErrRelayLogFileSizeSmaller.Generate(latestFilePath)
+				return
+			case cmp > 0:
+				updatePathCh <- latestFilePath
 				return
-			case event, ok := <-watcher2.Events:
-				if !ok {
-					result <- watchResult{
-						updatePath: "",
-						err:        terror.ErrWatcherChanClosed.Generate("events", dir),
-					}
+			default:
+				// current watched file size have no change means that no new writes have beed made
+				// our relay meat file will be updated immediately after receive the rotate event
+				// although we can't ensure the binlog file name in meta is the next one we expected
+				// but if we return a different filename with latestFile,the outer logic (parseDirAsPossible) will find the right one
+				metaFile, err := os.Open(filepath.Join(dir, utils.MetaFilename))
+				if err != nil {
+					errCh <- terror.Annotatef(err, "open metaFile from %s in dir %s", utils.MetaFilename, dir)
 					return
 				}
-				log.L().Debug("watcher receive event", zap.Reflect("event", event))
-				if event.IsDirEvent() {
-					log.L().Debug("skip watched event for directory", zap.Reflect("event", event))
-					continue
-				} else if !event.HasOps(watcher.Modify, watcher.Create) {
-					log.L().Debug("skip uninterested event", zap.Stringer("operation", event.Op), zap.String("path", event.Path))
-					continue
-				}
-				baseName := filepath.Base(event.Path)
-				if !binlog.VerifyFilename(baseName) {
-					log.L().Debug("skip watcher event for invalid relay log file", zap.Reflect("event", event))
-					continue // not valid binlog created, updated
+				defer metaFile.Close()
+				meta := &Meta{}
+				_, err = toml.DecodeReader(metaFile, meta)
+				if err != nil {
+					errCh <- terror.Annotate(err, "decode meta toml faild")
+					return
 				}
-				result <- watchResult{
-					updatePath: event.Path,
-					err:        nil,
+				if meta.BinLogName != latestFile {
+					nextFilePath := filepath.Join(dir, meta.BinLogName)
+					log.L().Info("newer relay log file is already generated, start parse from it",
+						zap.String("now file", latestFilePath),
+						zap.String("new file", nextFilePath))
+					updatePathCh <- nextFilePath
```

Comment:
check previous file doesn't have new content

Reply:
Can you please tell me the reason why I need to do this check

We check the current file size in each for loop first.
and only check the contents of the meta file when the file size is constant

The previous file in the current meta file is not necessarily the current file.
So it might not make sense to check if the size of the previous file has changed?

Reply:
after the check, new content may be written and then new relay log file generated